### ISSUE TITLE
native shortcode support

### ISSUE THREAD

**Issue by tablatronix** [GetSimpleCMS/GetSimpleCMS#557]
Add native shortcode support, similar to dynpages plugin, for insertion of components/snippets etc.

Add editor plugin for inserting registered shortcodes
Add filter to remove unused shortcodes before output

refs
#507

**Comment by tablatronix** [comment]
I already have code for this I might add it to core or as a core plugin similar to dynpages but to establish a gs standard.

The shortcodes will require several pieces.

Primary regex parser or Dom parser perhaps both.

Either way the element or string will contain several fields.
And methods to easily register them as callbacks like actions.

 * code id which will probably default to plugin Id
 * Action which is just an argument 
 * Additional arguments

**Comment by dimayakovlev** [comment]
@tablatronix It is very interesting for me.

**Comment by tablatronix** [comment]
I was just looking at word presses short code functionality and what I have already written is pretty much the same thing, so Our short does will likely be implemented the same if not similarly. Our parser is probably custom however.

**Comment by n00dles** [comment]
I already did the Wordpress shortcodes ages ago.. 8) 

http://get-simple.info/extend/plugin/gs-shortcodes/493/

**Comment by tablatronix** [comment]
hmm i never actually looked at your implementation, I didn't realize you used portable shortcode support. I can probably pull that into core, im sure it meets any criteria we need.

Is this fast, I do not think it matches tag patters first then process them, i think it tries to match all registered shortcodes right off the bat if i read the code right, else it might be a performance hit with all those | or's. This might be because it supports nested shortcodes, hmm

It seems to support wrapping container shortcodes using open closing tags ? That is pretty neat if works.

Does it support flags in addition to key:values, can the preset arg attrs be optional , so you can get all attrs ? Flags would be [code ON] [code 12] [code name=test TRUE] [code USESPECIFICENUMERATATIONFLAG]

**Comment by n00dles** [comment]
yeah its just a port of the WP one, and is mostly a copy paste of the functions although with some changes to fit into GS. 

The regex is slightly changed too from the WP one ( i think you helped me out with it back in the day) to better remove spaces/<p> tags etc...

It can support flags/unnamed vars and they will have a positional number in the $atts array

whatever is supported here (https://codex.wordpress.org/Shortcode_API) should be good in my version.

**Comment by tablatronix** [comment]
I'll have to look again and add in any enhancements I made to my regex and probably rewrite the. As inline wth comments.

My parser has the ability to pass in default arguments as functions not sure if it's needed or not but it might be interesting .


I do things like this.

'''
$allow arts = array(
  'depth' => create_function('$input', 'return filter_var(intval($input), FILTER_VALIDATE_INT, array(1,6));' ),
  'class' => create_function('$input', 'return filter_var($input, FILTER_SANITIZE_STRING, FILTER_FLAG_STRIP_LOW);'),
  'asc'   => create_function('$input', 'return true;'),
  'abs'   => create_function('$input', 'return true;'),
  'ul'    => create_function('$input', 'return true;'),
  'ol'    => create_function('$input', 'return true;'),
  'notoc' => create_function('$input', 'return true;'),
  'debug' => create_function('$input', 'return true;'),
  'demo'  => create_function('$input', 'return true;'),
);
'''
